{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                  Substitution_Rate\n",
      "Sequence_ID                                        \n",
      "Alphacoronavirus_Bat-CoV/P                 0.733333\n",
      "Bat_Coronavirus                            0.750000\n",
      "Bat_Coronavirus_BM48-31                    0.578947\n",
      "Beluga_Whale_Coronavirus_SW1               0.690476\n",
      "Betacoronavirus_England_1                  0.437500\n",
      "Betacoronavirus_HKU24_Strain_HKU           0.282051\n",
      "Bovine_Coronavirus_Isolate_BCoV-           0.023810\n",
      "Canada_Goose_Coronavirus_Strain_           0.576923\n",
      "Common-moorhen_Coronavirus_HKU21           0.809524\n",
      "Duck_Coronavirus_Isolate_DK/GD/2           0.000000\n",
      "Feline_Infectious_Peritonitis_Vi           0.547619\n",
      "Ferret_Coronavirus_Isolate_FRCoV           0.558824\n",
      "Isolate_Camel_Aplhacoronavirus             0.738095\n",
      "Lucheng_Rn_Rat_Coronavirus_Isola           0.560976\n",
      "Magpie-robin_Coronavirus_HKU18             0.738095\n",
      "Middle_East_Coronavirus_Isolate_           0.424242\n",
      "Mink_Coronavirus_Strain_WD1127             0.529412\n",
      "Munia_Coronavirus_HKU13-3514               0.738095\n",
      "Night-heron_Coronavirus_HKU19              0.761905\n",
      "Porcine_Coronavirus_HKU15_Strain           0.684211\n",
      "Rabbit_Coronavirus_HKU14                   0.097561\n",
      "Rodent_Coronavirus_Isolate_RtMru           0.317073\n",
      "Shrew_Coronavirus_Isolate_Shrew-           0.738095\n",
      "Sparrow_Coronavirus_HKU17                  0.611111\n",
      "Swine_Enteric_Coronavirus_Strain           0.487179\n",
      "Thrush_Coronavirus_HKU12-600               0.560976\n",
      "Turkey_coronavirus                         0.531250\n",
      "White-eye_Coronavirus_HKU16                0.666667\n",
      "Wigeon_Coronavirus_HKU20                   0.527778\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-335-a95b7039988f>:232: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.\n",
      "  sub_averages = strains['Sequence_ID', \"Substitution_Rate\"].mean()\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from pandas import read_csv\n",
    "\n",
    "import csv\n",
    "\n",
    "class Phylogeny_Data():\n",
    "    \"\"\"This class hold information and lists of strain objects\n",
    "    To represent the strains we are using in our phylogeny\"\"\"\n",
    "\n",
    "    names = []\n",
    "    \n",
    "    ##Holds an aligned sequence file\n",
    "    alignment_file = \"\"\n",
    "    \n",
    "    strains = []\n",
    "    \n",
    "    def __init__(self):\n",
    "        \"\"\"Creates a Phylogeny_Data obj\n",
    "        Takes in self\"\"\"\n",
    "    \n",
    "        alignment_file = \"\"\n",
    "        names = []\n",
    "    \n",
    "    def get_names(self):\n",
    "        \"\"\"Returns a list of names based on the current alignment_file\n",
    "        Takes in self\"\"\"\n",
    "    \n",
    "        names = []\n",
    "    \n",
    "        file = open(self.alignment_file)\n",
    "        \n",
    "        for line in file:\n",
    "\n",
    "            ##Split up into list of words\n",
    "            words = line.split()\n",
    "            \n",
    "            if len(words) == 0:\n",
    "                self.names = names\n",
    "                file.close()\n",
    "                return names\n",
    "            else:\n",
    "                ##take first word\n",
    "                name = words[0]\n",
    "                ##add to list\n",
    "                names.append(name)\n",
    "                \n",
    "    def get_strains(self):\n",
    "        \"\"\"Returns a list of strain objects from our alignment data\n",
    "        Takes in self\"\"\"\n",
    "      \n",
    "        for self.name in self.names:\n",
    "\n",
    "            genome = Strain()\n",
    "            genome.name = self.name\n",
    "            genome.sequences = []\n",
    "            \n",
    "            file = open(self.alignment_file)\n",
    "        \n",
    "            for line in file:\n",
    "                ##Split up into list of words\n",
    "                words = line.split()\n",
    "            \n",
    "                if len(words) == 0:\n",
    "                    continue\n",
    "                elif words[0] == self.name:\n",
    "                    genome.sequences.append(words[1])\n",
    "                elif words[0] == \"end\":\n",
    "                    break\n",
    "                    \n",
    "            self.strains.append(genome)\n",
    "    \n",
    "        file.close()\n",
    "        \n",
    "        return self.strains\n",
    "    \n",
    "    def compare_seq(self, selected_strain, selected_sequence):\n",
    "        \"\"\"function takes in an object strain, and an int sequence (out of 519 sequences)\n",
    "        compares selected sequence of strain to all other aligned sequences of other strains\n",
    "        example parameter (Bat_Coronavirus, 342)\n",
    "        return: a dict holding whats being compared to base as the key and its substitution rate as the value\"\"\"\n",
    "        \"\"\"example dict\n",
    "        'Beluga_coronavirus_sequence_342', '3/60 substitutions took place with Bat_coronavirus_sequence_342'\n",
    "        'Feline_coronavirus_sequence_233', '12/60 substitutions took place with Bat_coronavirus_sequence_233'\n",
    "        \"\"\"\n",
    "        # if counting deletions\n",
    "        total_comparisons = 60\n",
    "    \n",
    "        # if ignoring deletions, counter reset after each comparison\n",
    "        non_deletion_comparisons = 0\n",
    "    \n",
    "        # number of nucleotide matches, reset after each comparison\n",
    "        matches = 0\n",
    "    \n",
    "        # dict to hold compared strain and sequence as key, sub rate with selected strain as value\n",
    "        substitution_rates = []\n",
    "    \n",
    "        # list to hold all aligned sequences of nth alignment to be compared relative to chosen strain\n",
    "        alignments = []\n",
    "        \n",
    "        # String to hold all matches within a comparison ('*' means that the nucleotides did not match)\n",
    "        matches_as_string = \"\"\n",
    "    \n",
    "        # grabbing selected sequence as bases of comparison\n",
    "        selected_sequence_as_string = selected_strain.sequences[selected_sequence]\n",
    "        \n",
    "        # track what sequence name we are on\n",
    "        strain_names = []\n",
    "\n",
    "        \"\"\"adding nth sequence alignment for all 30 strains to list including selected sequence. we will deal with \n",
    "        avoiding comparing sequence to itself below\"\"\"\n",
    "        for element in self.strains:\n",
    "            if selected_sequence > len(element.sequences) - 1:\n",
    "                break\n",
    "                \n",
    "            if selected_strain.name == element.name:\n",
    "                continue\n",
    "                \n",
    "            strain_names.append(element.name)\n",
    "            \n",
    "            alignments.append(element.sequences[selected_sequence])\n",
    "\n",
    "        \"\"\"iterating thru sequences in list, then iterating thru each sequence and doing comparisons of nucleotides to \n",
    "        selected sequence nucleotides, count up matches and nucleotide comparisons, add results to substitution rate \n",
    "        dict, not comparing strain sequence to itself\"\"\"\n",
    "        \n",
    "        #set seq count\n",
    "        seq_count = 0\n",
    "    \n",
    "        # iterating thru 30 selected sequences\n",
    "        for seq in alignments:\n",
    "            \n",
    "            # reset count\n",
    "            count = 0\n",
    "            # reset match string\n",
    "            matches_as_string = \"\"\n",
    "            \n",
    "            #print(seq)\n",
    "            #print(selected_sequence_as_string)\n",
    "            \n",
    "            # do not compare if sequences are of the same strain\n",
    "            if not seq == selected_sequence_as_string:\n",
    "                # iterating through 60 char sequence, ignoring deletions, adding up matches\n",
    "                for nucleotide in seq:\n",
    "                    \n",
    "                    ##print(str(nucleotide) + \",\" + str(selected_sequence_as_string[count]))\n",
    "                    \n",
    "                    if (count > len(selected_sequence_as_string) - 1):\n",
    "                        break\n",
    "                    \n",
    "                    # check for deletion, add one if no deletion\n",
    "                    if not nucleotide == \"-\" and not selected_sequence_as_string[count] == \"-\":\n",
    "                        non_deletion_comparisons += 1\n",
    "                        \n",
    "                        # check for match, add to count if so\n",
    "                        if nucleotide == selected_sequence_as_string[count]:\n",
    "                            matches += 1\n",
    "                            \n",
    "                            #print(str(nucleotide) + \",\" + str(selected_sequence_as_string[count]))\n",
    "                            \n",
    "                            ##add to matches string\n",
    "                            matches_as_string += selected_sequence_as_string[count]\n",
    "                        else:\n",
    "                            matches_as_string += \"*\"\n",
    "                            \n",
    "                        # if deletion detected, continue\n",
    "                    else:\n",
    "                        ##add to matched string\n",
    "                        matches_as_string += \"-\"\n",
    "                        \n",
    "                        # add to count\n",
    "                        count += 1\n",
    "                        \n",
    "                        continue\n",
    "                            \n",
    "                    # add to count\n",
    "                    count += 1\n",
    "                    \n",
    "                \"\"\"after each individual sequence comparison is made, add sub rate to dict, and reset num of match and \n",
    "                num of comparison counter for new sequence comparison\"\"\"\n",
    "                non_matches = non_deletion_comparisons - matches\n",
    "                non_matches_as_fraction = str(non_matches) + \"/\" + str(non_deletion_comparisons)\n",
    "                \n",
    "                # testing results\n",
    "                #print(matches)\n",
    "                #print(non_deletion_comparisons)\n",
    "                #print(non_matches)\n",
    "                #print(non_matches_as_fraction)\n",
    "                #print(matches_as_string)\n",
    "            \n",
    "                ## ensure that we don't divide by zero\n",
    "                if non_deletion_comparisons == 0:\n",
    "                    sub_rate = 0.00\n",
    "                else:\n",
    "                    sub_rate = non_matches / non_deletion_comparisons\n",
    "                \n",
    "                value = sub_rate\n",
    "                \n",
    "                data = {\"Sequence_ID\": str(strain_names[seq_count]), \n",
    "                        \"Substitution_Rate\" : value, \"Matched_Nucleotides\" : matches_as_string}\n",
    "                \n",
    "                substitution_rates.append(data)\n",
    "                                          \n",
    "                matches = 0\n",
    "                non_deletion_comparisons = 0\n",
    "                \n",
    "                seq_count += 1\n",
    "            \n",
    "            else:\n",
    "                \n",
    "                seq_count += 1\n",
    "                continue\n",
    "        # after all 29 strain sequences are compared to the selected strain sequence, return dict with sub rate info\n",
    "        return substitution_rates\n",
    "        \n",
    "    def create_table(self, sub_rates, strain):\n",
    "        \"\"\"Returns nothing, displays a table of the substitution analysis results\n",
    "        Takes in a dict of seqs and associated sub rates, and a strain name\"\"\"\n",
    "        \n",
    "        count = 0 \n",
    "            \n",
    "        #Load the text version of the table (a csv file) into python using pandas\n",
    "            \n",
    "        feature_table = pd.read_csv(\"export.csv\")\n",
    "            \n",
    "        count += 1\n",
    "        \n",
    "        feature_table.columns = ['Sequence_ID', 'Substitution_Rate', 'Matched_Nucleotides']\n",
    "\n",
    "        strains = feature_table.groupby('Sequence_ID')\n",
    "\n",
    "        sub_averages = strains['Sequence_ID', \"Substitution_Rate\"].mean()\n",
    "\n",
    "        print(sub_averages)\n",
    "\n",
    "        \n",
    "    def create_data(self, strain_list, index):\n",
    "        \"\"\"returns an int reporting how many .csv files were made containing all \n",
    "        of the sequence comparisons for a given strain. The average is taken of all \n",
    "        of the strains sub_rates. Creates a .csv filed titled \"export_all\" containing all data\n",
    "        Takes in self, a strain obj list, and an int of the index of the strain we want to\n",
    "        compare all others to\"\"\"\n",
    "        \n",
    "        counter = 0\n",
    "        \n",
    "        data = []\n",
    "        \n",
    "        for seq in strain_list[index].sequences:\n",
    "            # compare every seq of chosen strain\n",
    "            data.extend(self.compare_seq(strain_list[index], counter))\n",
    "            \n",
    "            counter += 1\n",
    "            \n",
    "            \n",
    "        ##explictly name columns\n",
    "        col_name =[\"Sequence_ID\",\"Substitution_Rate\", \"Matched_Nucleotides\"]\n",
    "        \n",
    "        # Write to exportfile based on counter\n",
    "        with open(\"export_all.csv\", 'w') as csvFile:\n",
    "            wr = csv.DictWriter(csvFile, fieldnames = col_name, lineterminator = '\\n')\n",
    "            wr.writeheader()\n",
    "            for element in data:\n",
    "                wr.writerow(element)\n",
    "                  \n",
    "        return counter - 1\n",
    "    \n",
    "    def create_one_data_set(self, strain_list, index, seq_index):\n",
    "        \"\"\"returns nothing, creates a file titled export containing the info for\n",
    "        one sequence comparison of a chosen base strain\n",
    "        Takes in self, a list of current strains, the \"\"\"\n",
    "        \n",
    "        data = (self.compare_seq(strain_list[index], seq_index))\n",
    "        \n",
    "        ##explictly name columns\n",
    "        col_name =[\"Sequence_ID\",\"Substitution_Rate\", \"Matched_Nucleotides\"]\n",
    "        \n",
    "        # Write to exportfile based on counter\n",
    "        with open(\"export.csv\", 'w') as csvFile:\n",
    "            wr = csv.DictWriter(csvFile, fieldnames = col_name, lineterminator = '\\n')\n",
    "            wr.writeheader()\n",
    "            for element in data:\n",
    "                wr.writerow(element)\n",
    "        \n",
    "    \n",
    "              \n",
    "class Strain():\n",
    "    \"\"\"This class hold info about the strain we are analyzing\"\"\"\n",
    "\n",
    "    sequences = []\n",
    "    ##A list to hold sequences\n",
    "\n",
    "    name = \"\"\n",
    "    ##String to hold strain name\n",
    "\n",
    "    def __init__(dictionary, seq_name):\n",
    "    ##Constructs a strain obj using the given parameters\n",
    "        sequences = dictionary\n",
    "        name = seq_name\n",
    "        \n",
    "    def __init__(self):\n",
    "    ##No args constructor\n",
    "        name = \"\"\n",
    "        sequences = None\n",
    "        \n",
    "    def to_string(self):\n",
    "    ## returns a string containing obj info\n",
    "    \n",
    "        ## name line \n",
    "        string = \">\" + str(self.name) + \"\\n\"\n",
    "        \n",
    "        for self.seq in self.sequences:\n",
    "            string += str(self.seq + \"\\n\")\n",
    "            \n",
    "        return string\n",
    "    \n",
    "## Driver code\n",
    "    \n",
    "phylogeny = Phylogeny_Data()\n",
    "\n",
    "phylogeny.alignment_file = \"Alignments.txt\"\n",
    "\n",
    "phylogeny.get_names()\n",
    "\n",
    "curr_strains = phylogeny.get_strains() \n",
    "\n",
    "output = open(\"Output.txt\", 'w')\n",
    "\n",
    "for strain in curr_strains:\n",
    "    output.write((strain.to_string()) + \"\\n\")\n",
    "\n",
    "output.close()\n",
    "\n",
    "output = open(\"Output_Comp.txt\", 'w')\n",
    "\n",
    "data = phylogeny.compare_seq(curr_strains[10], 0)\n",
    "\n",
    "## save our current compared strain data as .txt file\n",
    "output.write(str(data))\n",
    "\n",
    "## Create data sets as .csv files\n",
    "phylogeny.create_data(curr_strains, 10)\n",
    "\n",
    "output.close()\n",
    "\n",
    "phylogeny.create_one_data_set(curr_strains, 10, 0)\n",
    "\n",
    "phylogeny.create_table(data, curr_strains[10])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
