{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Bat_Coronavirus', 'Beluga_Whale_Coronavirus_SW1', 'Bat_Coronavirus_BM48-31', 'Night-heron_Coronavirus_HKU19', 'Wigeon_Coronavirus_HKU20', 'Common-moorhen_Coronavirus_HKU21', 'Betacoronavirus_HKU24_Strain_HKU', 'Rodent_Coronavirus_Isolate_RtMru', 'Rabbit_Coronavirus_HKU14', 'Bovine_Coronavirus_Isolate_BCoV-', 'Human_Coronavirus_OC43_Strain_AT', 'Canada_Goose_Coronavirus_Strain_', 'Duck_Coronavirus_Isolate_DK/GD/2', 'Turkey_coronavirus', 'Betacoronavirus_England_1', 'Middle_East_Coronavirus_Isolate_', 'Thrush_Coronavirus_HKU12-600', 'Magpie-robin_Coronavirus_HKU18', 'Munia_Coronavirus_HKU13-3514', 'White-eye_Coronavirus_HKU16', 'Sparrow_Coronavirus_HKU17', 'Porcine_Coronavirus_HKU15_Strain', 'Shrew_Coronavirus_Isolate_Shrew-', 'Lucheng_Rn_Rat_Coronavirus_Isola', 'Swine_Enteric_Coronavirus_Strain', 'Feline_Infectious_Peritonitis_Vi', 'Mink_Coronavirus_Strain_WD1127', 'Ferret_Coronavirus_Isolate_FRCoV', 'Alphacoronavirus_Bat-CoV/P', 'Isolate_Camel_Aplhacoronavirus']\n"
     ]
    }
   ],
   "source": [
    "class Phylogeny_Data():\n",
    "##This class hold information and lists of strain objects\n",
    "##To represent the strains we are using in our phylogeny\n",
    "\n",
    "    names = []\n",
    "    \n",
    "    ##Holds an aligned sequence file\n",
    "    alignment_file = \"\"\n",
    "    \n",
    "    strains = []\n",
    "    \n",
    "    def __init__(self):\n",
    "    ##Creates a Phylogeny_Data obj\n",
    "        alignment_file = \"\"\n",
    "        names = []\n",
    "    \n",
    "    def get_names(self):\n",
    "    ##Returns a list of names based on the current alignment_file\n",
    "    \n",
    "        names = []\n",
    "    \n",
    "        file = open(self.alignment_file)\n",
    "        \n",
    "        for line in file:\n",
    "\n",
    "            ##Split up into list of words\n",
    "            words = line.split()\n",
    "            \n",
    "            if len(words) == 0:\n",
    "                self.names = names\n",
    "                file.close()\n",
    "                return names\n",
    "            else:\n",
    "                ##take first word\n",
    "                name = words[0]\n",
    "                ##add to list\n",
    "                names.append(name)\n",
    "                \n",
    "    def get_strains(self):\n",
    "    ##Returns a list of strain objects from our alignment data\n",
    "      \n",
    "        for self.name in self.names:\n",
    "\n",
    "            genome = Strain()\n",
    "            genome.name = self.name\n",
    "            \n",
    "            file = open(self.alignment_file)\n",
    "        \n",
    "            for line in file:\n",
    "                ##Split up into list of words\n",
    "                words = line.split()\n",
    "            \n",
    "                if len(words) == 0:\n",
    "                    continue\n",
    "                elif words[0] == self.name:\n",
    "                    genome.sequences.append(words[1])\n",
    "                elif words[0] == \"end\":\n",
    "                    break\n",
    "                    \n",
    "            self.strains.append(genome)\n",
    "    \n",
    "        file.close()\n",
    "        return self.strains\n",
    "            \n",
    "class Strain():\n",
    "##This class hold info about the strain we are analyzing\n",
    "\n",
    "    sequences = []\n",
    "    ##A list to hold sequences\n",
    "\n",
    "    name = \"\"\n",
    "    ##String to hold strain name\n",
    "\n",
    "    def __init__(dictionary, seq_name):\n",
    "    ##Constructs a strain obj using the given parameters\n",
    "        sequences = dictionary\n",
    "        name = seq_name\n",
    "        \n",
    "    def __init__(self):\n",
    "    ##No args constructor\n",
    "        name = \"\"\n",
    "        sequences = None\n",
    "        \n",
    "    def to_string(self):\n",
    "    ## returns a string containing obj info\n",
    "    \n",
    "        ## name line \n",
    "        string = \">\" + str(self.name) + \"\\n\"\n",
    "        \n",
    "        for self.seq in self.sequences:\n",
    "            string += str(self.seq + \"\\n\")\n",
    "            \n",
    "        return string\n",
    "    \n",
    "phylogeny = Phylogeny_Data()\n",
    "\n",
    "phylogeny.alignment_file = \"Alignments.txt\"\n",
    "\n",
    "print(str(phylogeny.get_names()))\n",
    "\n",
    "curr_strains = phylogeny.get_strains() \n",
    "\n",
    "output = open(\"Output.txt\", 'w')\n",
    "\n",
    "for strain in curr_strains:\n",
    "    output.write((strain.to_string()) + \"\\n\")\n",
    "\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
